contrib/native/client/src/clientlib/saslAuthenticatorImpl.cpp

```diff
@@ -0,0 +1,206 @@
+            }
+
+            sasl_client_init(NULL);
+            { // for debugging purposes
+                const char **mechanisms = sasl_global_listmech();
+                int i = 0;
+                DRILL_MT_LOG(DRILL_LOG(LOG_TRACE) << "SASL mechanisms available on client: " << std::endl;)
+                while (mechanisms[i] != NULL) {
+                    DRILL_MT_LOG(DRILL_LOG(LOG_TRACE) << i << " : " << mechanisms[i] << std::endl;)
+                    i++;
+                }
+            }
+            s_initialized = true;
+        }
+    }
+}
+
+SaslAuthenticatorImpl::~SaslAuthenticatorImpl() {
+    if (m_secret) {
+        free(m_secret);
+    }
+    // may be used to negotiated security layers before disposing in the future
+    if (m_pConnection) {
+        sasl_dispose(&m_pConnection);
+    }
+    m_pConnection = NULL;
+}
+
+typedef int (*sasl_callback_proc_t)(void); // see sasl_callback_ft
+
+int SaslAuthenticatorImpl::userNameCallback(void *context, int id, const char **result, unsigned *len) {
+    const std::string* const username = static_cast<const std::string* const>(context);
+
+    if ((SASL_CB_USER == id || SASL_CB_AUTHNAME == id)
+        && username != NULL) {
+        *result = username->c_str();
+        // *len = (unsigned int) username->length();
+    }
+    return SASL_OK;
+}
+
+int SaslAuthenticatorImpl::passwordCallback(sasl_conn_t *conn, void *context, int id, sasl_secret_t **psecret) {
+    const SaslAuthenticatorImpl* const authenticator = static_cast<const SaslAuthenticatorImpl* const>(context);
+
+    if (SASL_CB_PASS == id) {
+        const std::string password = authenticator->m_password;
+        const size_t length = password.length();
+        authenticator->m_secret->len = length;
+        std::memcpy(authenticator->m_secret->data, password.c_str(), length);
+        *psecret = authenticator->m_secret;
+    }
+    return SASL_OK;
+}
+
+int SaslAuthenticatorImpl::init(const std::vector<std::string>& mechanisms, exec::shared::SaslMessage& response) {
+    // find and set parameters
+    std::string authMechanismToUse;
+    std::string serviceName;
+    std::string serviceHost;
+    for (size_t i = 0; i < m_properties->size(); i++) {
+        const std::string key = m_properties->keyAt(i);
+        const std::string value = m_properties->valueAt(i);
+
+        if (USERPROP_SERVICE_HOST == key) {
+            serviceHost = value;
+        } else if (USERPROP_SERVICE_NAME == key) {
+            serviceName = value;
+        } else if (USERPROP_PASSWORD == key) {
+            m_password = value;
+            m_secret = (sasl_secret_t *) malloc(sizeof(sasl_secret_t) + m_password.length());
+            authMechanismToUse = PLAIN_NAME;
+        } else if (USERPROP_USERNAME == key) {
+            m_username = value;
+        } else if (USERPROP_AUTH_MECHANISM == key) {
+            authMechanismToUse = value;
+        }
+    }
+    if (authMechanismToUse.empty()) return SASL_NOMECH;
```

Comment:
For cpp client we always expect **authMechanism** to be mentioned as parameter ? Since on java side we look for other Kerberos/Plain mechanism related parameters and use it in case a particular mechanism is not specified on client side.